### ISSUE TITLE
SimplExpr Lifts Table key calls into control declaration

### ISSUE THREAD

**Issue by rudynicolop** [verified-network-toolchain/petr4#391]
In `P4light/Transformations/SimplExpr.v` function calls are moved out of expressions to make them stateless.

This may be problematic for table keys, where function calls are moved out of the key expression into a variable declaration in the control block.

[The P4 Spec forbids functions from being called at the top-level of a control block](https://p4.org/p4-spec/docs/P4-16-v-1.2.3.html#sec-calling-restrictions).

I'm not sure the best way to resolve this, table syntax in P4light and P4cub may need to be adjusted to have statements for instance.

**Comment by QinshiWang** [comment]
Also, moving function calls in table keys to the top level of a control block is incorrect (w.r.t. when the functions are called). If we need to resolve this problem, I suppose we need to add a `statement` member to the table constructor.

**Comment by hackedy** [comment]
I think the solution is to create a fresh variable for the table's key and assign the original key expression to that variable right before wherever the table gets applied.

For example if we have
```
bit<8> f(in bit<8> arg) { ... }
table t {
  key = f(e);
}
apply {
  t.apply();
}
```
we could safely lift it to 
```
bit<8> fresh_var;
table t {
  key = fresh_var;
}
apply {
  fresh_var = f(e);
  t.apply();
}
```

**Comment by QinshiWang** [comment]
@hackedy It breaks the original module boundary of tables. I don't really like it.

**Comment by hackedy** [comment]
Yeah, it's also not going to work if variables in `e` are shadowed between the declaration of the table and its use. For example
```
int<8> x = 0;
table t {
  key = f(x)
}
int<8> x = 1;
apply {
  t.apply()
}
```

**Comment by QinshiWang** [comment]
Yeah. Variable shadowing may be avoidable by locator. But the current design is generating locators after simplifying expressions, so the latter does not need to handle locators.

**Comment by hackedy** [comment]
@ericthewry says he has a partial fix for this on his deadline branch which he can extend to a full fix before merging.

**Comment by hackedy** [comment]
> @hackedy It breaks the original module boundary of tables. I don't really like it.

The current approach violates evaluation order for table keys. There is no module boundary around a table, they act more like macros than functions. My proposed fix preserves evaluation order.

**Comment by QinshiWang** [comment]
A table has its boundary as a function. Key evaluation should happen inside `table.apply()` in order to reason about it.
